xarray/core/indexing.py

```diff
@@ -519,9 +525,12 @@ def shape(self) -> tuple[int, ...]:
                 shape.append(k.size)
         return tuple(shape)
 
-    def __array__(self, dtype=None):
+    def get_duck_array(self):
         array = as_indexable(self.array)
-        return np.asarray(array[self.key], dtype=None)
+        array = array[self.key]
+        if isinstance(array, ExplicitlyIndexed):
+            array = array.get_duck_array()
```

Comment:
Could this be replaced by just:
```suggestion
       array = array.get_duck_array()
```
?

I think the result of indexing an ExplicitlyIndexed array should probably always still be explicitly indexed?

Reply:
`array[self.key]` returns a numpy array when 
```
ipdb> self.array
<xarray.backends.netCDF4_.NetCDF4ArrayWrapper object at 0x7fce3d0ef500>
ipdb> self.key
BasicIndexer((slice(None, None, None),))
```

I traced it back to `class BackendArray(NdimSizeLenMixin, indexing.ExplicitlyIndexed)`. Is this a bug?

Reply:
Maybe? It's mostly a matter of having well-defined and type-stable internal interfaces.

Ideally all internal Xarray functionality would _either_ work for duck-arrays or `ExplicitlyIndexed` arrays, with no need for conditional logic dependent on error-prone runtime type checking.

To help nail this down, a type checker like mypy could potentially be quite helpful.

Reply:
This illustrates the problem
```
LazilyIndexedArray(
	array=_ElementwiseFunctionArray(
		LazilyIndexedArray(
			array=<xarray.backends.netCDF4_.NetCDF4ArrayWrapper object at 0x7f9a073a6740>, 
			key=BasicIndexer((slice(None, None, None),))), 
		func=functools.partial(
			<function decode_cf_datetime at 0x7f9a006b0af0>, units='days since 2000-01-01 00:00:00', calendar='proleptic_gregorian', use_cftime=None
			), dtype=dtype('<M8[ns]')), 
	key=BasicIndexer((slice(None, None, None),)))
)
```

This method is removing `LazilyIndexedArray`. When we do that, we sometimes have another ExplicitlyIndexed array  (_ElementwiseFunctionArray) and sometimes a BackendArray. We then apply `array[self.key]` which returns a ExplicitlyIndexed array for the former and forces a load from disk for the latter.

One way to solve this would be to return another ExplicitlyIndexed array from the `__getitem__` on the BackendArrays. This is currently forcing a load from disk:

https://github.com/pydata/xarray/blob/fbaf815af746a73b29c2b927ccb3aaf37f3a8721/xarray/backends/netCDF4_.py#L94-L97